api/src/main/java/net/kyori/adventure/text/width/PixelWidthSourceImpl.java

```diff
@@ -0,0 +1,368 @@
+    int length = 0;
+    for(final char c : string.toCharArray()) length += this.width(c, style, context);
+    return length;
+  }
+
+  @Override
+  public int width(final char c, final @NonNull Style style, final @NonNull CX context) {
+    return this.function.apply(context).applyAsInt(c, style);
+  }
+
+  @SuppressWarnings(value="unchecked")
+  @Override
+  public <CO extends Component> void addResolver(final @NonNull Class<CO> resolveFor, final @NonNull ComponentResolver<CO, CX> resolver) {
+
+    if(resolveFor.isAssignableFrom(TextComponent.class)) {
+      throw new UnsupportedOperationException("Can not add custom resolver for TextComponents");
+    } else if(resolveFor.isAssignableFrom(TranslatableComponent.class)) {
+      this.translatableResolvers.add((ComponentResolver<TranslatableComponent, CX>) resolver);
+      return;
+    } else if(resolveFor.isAssignableFrom(KeybindComponent.class)) {
+      this.keybindResolvers.add((ComponentResolver<KeybindComponent, CX>) resolver);
+      return;
+    } else if(resolveFor.isAssignableFrom(ScoreComponent.class)) {
+      this.scoreResolvers.add((ComponentResolver<ScoreComponent, CX>) resolver);
+      return;
+    } else if(resolveFor.isAssignableFrom(SelectorComponent.class)) {
+      this.selectorResolvers.add((ComponentResolver<SelectorComponent, CX>) resolver);
+      return;
+    } else if(resolveFor.isAssignableFrom(NBTComponent.class)) {
+      if(resolveFor.isAssignableFrom(BlockNBTComponent.class)) {
+        this.blockNBTResolvers.add((ComponentResolver<BlockNBTComponent, CX>) resolver);
+        return;
+      } else if(resolveFor.isAssignableFrom(EntityNBTComponent.class)) {
+        this.entityNBTResolvers.add((ComponentResolver<EntityNBTComponent, CX>) resolver);
+        return;
+      } else if(resolveFor.isAssignableFrom(StorageNBTComponent.class)) {
+        this.storageNBTResolvers.add((ComponentResolver<StorageNBTComponent, CX>) resolver);
+        return;
+      }
+    }
+
+    throw new UnsupportedOperationException("Invalid Component");
+
+  }
+
+  /**
+   * Serializes using the renderer.
+   *
+   * @param component the component
+   * @since 4.5.0
+   */
+  private Map<String, Style> serialize(final @NonNull Component component, final @NonNull CX context) {
+
+    final TextComponent rendered = this.renderer.render(component, context);
+
+    return new HashMap<>(this.flattenComponents(rendered));
+  }
+
+  /**
+   * Flattens a {@link TextComponent} and <em>all</em> it's {@link TextComponent} children,
+   * wherever they are, into a map where the content and style of each component is linked together.
+   */
+  private Map<String, Style> flattenComponents(final @NonNull TextComponent parent){
+    //By this point, we know that all components in this component tree are TextComponents,
+    //so some casting is acceptable
+    final Map<String, Style> map = new HashMap<>();
```

Comment:
`flattenComponents` should take a map that it adds to, rather than creating a new Map at each recursion